File: backend/core/medical_standards.py
# backend/core/medical_standards.py
"""
Standar medis terpusat untuk indikator rumah sakit
Berdasarkan standar Kemenkes RI dan best practices internasional
"""

class MedicalStandards:
    """Konstanta standar medis untuk indikator rumah sakit"""
    
    # BOR (Bed Occupancy Rate) Standards
    BOR_OPTIMAL_MIN = 60.0      # Minimum optimal BOR (%)
    BOR_OPTIMAL_MAX = 85.0      # Maximum optimal BOR (%)  
    BOR_CRITICAL_HIGH = 90.0    # Critical high threshold (%)
    BOR_WARNING_LOW = 50.0      # Warning low threshold (%)
    
    # LOS (Length of Stay) Standards  
    LOS_OPTIMAL_MIN = 6.0       # Minimum optimal LOS (days)
    LOS_OPTIMAL_MAX = 9.0       # Maximum optimal LOS (days)
    LOS_CRITICAL_HIGH = 12.0    # Critical high threshold (days)
    LOS_WARNING_LOW = 3.0       # Warning low threshold (days)
    
    # BTO (Bed Turnover) Standards
    BTO_OPTIMAL_MIN = 40.0      # Minimum optimal BTO (times/month)
    BTO_OPTIMAL_MAX = 50.0      # Maximum optimal BTO (times/month)
    BTO_CRITICAL_LOW = 30.0     # Critical low threshold
    BTO_WARNING_HIGH = 60.0     # Warning high threshold
    
    # TOI (Turn Over Interval) Standards  
    TOI_OPTIMAL_MIN = 1.0       # Minimum optimal TOI (days)
    TOI_OPTIMAL_MAX = 3.0       # Maximum optimal TOI (days)
    TOI_CRITICAL_HIGH = 5.0     # Critical high threshold (days)
    TOI_WARNING_LOW = 0.5       # Warning very low threshold
    
    @classmethod
    def evaluate_bor(cls, bor_value: float) -> dict:
        """Evaluasi status BOR berdasarkan standar medis"""
        if bor_value >= cls.BOR_CRITICAL_HIGH:
            return {
                "status": "critical", 
                "level": "danger",
                "message": f"BOR {bor_value:.1f}% sangat tinggi - Risiko overkapasitas",
                "recommendation": "Tambah kapasitas atau percepat discharge"
            }
        elif bor_value > cls.BOR_OPTIMAL_MAX:
            return {
                "status": "warning", 
                "level": "warning",
                "message": f"BOR {bor_value:.1f}% di atas optimal",
                "recommendation": "Monitor kapasitas dan siapkan rencana darurat"
            }
        elif bor_value >= cls.BOR_OPTIMAL_MIN:
            return {
                "status": "optimal", 
                "level": "success",
                "message": f"BOR {bor_value:.1f}% dalam rentang optimal",
                "recommendation": "Pertahankan tingkat utilisasi ini"
            }
        elif bor_value >= cls.BOR_WARNING_LOW:
            return {
                "status": "low", 
                "level": "info",
                "message": f"BOR {bor_value:.1f}% rendah",
                "recommendation": "Evaluasi strategi pemasaran/rujukan"
            }
        else:
            return {
                "status": "critical_low", 
                "level": "danger",
                "message": f"BOR {bor_value:.1f}% sangat rendah - Under-utilisasi",
                "recommendation": "Review strategi operasional dan marketing"
            }
    
    @classmethod 
    def evaluate_los(cls, los_value: float) -> dict:
        """Evaluasi status LOS berdasarkan standar medis"""
        if los_value >= cls.LOS_CRITICAL_HIGH:
            return {
                "status": "critical", 
                "level": "danger",
                "message": f"LOS {los_value:.1f} hari terlalu panjang",
                "recommendation": "Review protokol discharge dan case management"
            }
        elif los_value > cls.LOS_OPTIMAL_MAX:
            return {
                "status": "warning", 
                "level": "warning", 
                "message": f"LOS {los_value:.1f} hari di atas optimal",
                "recommendation": "Evaluasi efisiensi perawatan"
            }
        elif los_value >= cls.LOS_OPTIMAL_MIN:
            return {
                "status": "optimal", 
                "level": "success",
                "message": f"LOS {los_value:.1f} hari dalam rentang optimal",
                "recommendation": "Pertahankan kualitas perawatan"
            }
        elif los_value >= cls.LOS_WARNING_LOW:
            return {
                "status": "short", 
                "level": "info",
                "message": f"LOS {los_value:.1f} hari pendek",
                "recommendation": "Monitor kualitas outcome pasien"
            }
        else:
            return {
                "status": "critical_short", 
                "level": "warning",
                "message": f"LOS {los_value:.1f} hari sangat pendek",
                "recommendation": "Pastikan tidak ada discharge terlalu dini"
            }
    
    @classmethod
    def evaluate_bto(cls, bto_value: float) -> dict:
        """Evaluasi status BTO berdasarkan standar medis"""
        if bto_value >= cls.BTO_WARNING_HIGH:
            return {
                "status": "high", 
                "level": "warning",
                "message": f"BTO {bto_value:.1f} sangat tinggi",
                "recommendation": "Evaluasi kualitas perawatan dan patient satisfaction"
            }
        elif bto_value >= cls.BTO_OPTIMAL_MIN:
            return {
                "status": "optimal", 
                "level": "success",
                "message": f"BTO {bto_value:.1f} dalam rentang optimal",
                "recommendation": "Pertahankan efisiensi turnover"
            }
        elif bto_value >= cls.BTO_CRITICAL_LOW:
            return {
                "status": "low", 
                "level": "info",
                "message": f"BTO {bto_value:.1f} rendah",
                "recommendation": "Evaluasi strategi admission dan discharge"
            }
        else:
            return {
                "status": "critical_low", 
                "level": "danger",
                "message": f"BTO {bto_value:.1f} sangat rendah",
                "recommendation": "Review komprehensif proses operasional"
            }
    
    @classmethod
    def get_all_thresholds(cls) -> dict:
        """Ambil semua threshold untuk export ke frontend"""
        return {
            "BOR": {
                "optimal_min": cls.BOR_OPTIMAL_MIN,
                "optimal_max": cls.BOR_OPTIMAL_MAX,
                "critical_high": cls.BOR_CRITICAL_HIGH,
                "warning_low": cls.BOR_WARNING_LOW
            },
            "LOS": {
                "optimal_min": cls.LOS_OPTIMAL_MIN,
                "optimal_max": cls.LOS_OPTIMAL_MAX,
                "critical_high": cls.LOS_CRITICAL_HIGH,
                "warning_low": cls.LOS_WARNING_LOW
            },
            "BTO": {
                "optimal_min": cls.BTO_OPTIMAL_MIN,
                "optimal_max": cls.BTO_OPTIMAL_MAX,
                "critical_low": cls.BTO_CRITICAL_LOW,
                "warning_high": cls.BTO_WARNING_HIGH
            },
            "TOI": {
                "optimal_min": cls.TOI_OPTIMAL_MIN,
                "optimal_max": cls.TOI_OPTIMAL_MAX,
                "critical_high": cls.TOI_CRITICAL_HIGH,
                "warning_low": cls.TOI_WARNING_LOW
            }
        }

File: frontend/src/utils/medicalStandards.ts
// Medical Standards Constants - Sinkron dengan backend
export interface MedicalThresholds {
  optimal_min: number;
  optimal_max: number;
  critical_high?: number;
  critical_low?: number;
  warning_low?: number;
  warning_high?: number;
}

export interface MedicalStandardsType {
  BOR: MedicalThresholds;
  LOS: MedicalThresholds;
  BTO: MedicalThresholds;
  TOI: MedicalThresholds;
}

// Default standards (akan di-sync dengan backend via API)
export const DEFAULT_MEDICAL_STANDARDS: MedicalStandardsType = {
  BOR: {
    optimal_min: 60.0,
    optimal_max: 85.0,
    critical_high: 90.0,
    warning_low: 50.0
  },
  LOS: {
    optimal_min: 6.0,
    optimal_max: 9.0,
    critical_high: 12.0,
    warning_low: 3.0
  },
  BTO: {
    optimal_min: 40.0,
    optimal_max: 50.0,
    critical_low: 30.0,
    warning_high: 60.0
  },
  TOI: {
    optimal_min: 1.0,
    optimal_max: 3.0,
    critical_high: 5.0,
    warning_low: 0.5
  }
};

export type EvaluationStatus = 'optimal' | 'warning' | 'critical' | 'low' | 'high';
export type EvaluationLevel = 'success' | 'warning' | 'danger' | 'info';

export interface EvaluationResult {
  status: EvaluationStatus;
  level: EvaluationLevel;
  message: string;
  recommendation: string;
}

export class MedicalStandardsEvaluator {
  private standards: MedicalStandardsType;

  constructor(standards: MedicalStandardsType = DEFAULT_MEDICAL_STANDARDS) {
    this.standards = standards;
  }

  updateStandards(newStandards: MedicalStandardsType) {
    this.standards = newStandards;
  }

  evaluateBOR(borValue: number): EvaluationResult {
    const bor = this.standards.BOR;
    
    if (borValue >= (bor.critical_high || 90)) {
      return {
        status: 'critical',
        level: 'danger',
        message: `BOR ${borValue.toFixed(1)}% sangat tinggi - Risiko overkapasitas`,
        recommendation: 'Tambah kapasitas atau percepat discharge'
      };
    } else if (borValue > bor.optimal_max) {
      return {
        status: 'warning',
        level: 'warning',
        message: `BOR ${borValue.toFixed(1)}% di atas optimal`,
        recommendation: 'Monitor kapasitas dan siapkan rencana darurat'
      };
    } else if (borValue >= bor.optimal_min) {
      return {
        status: 'optimal',
        level: 'success',
        message: `BOR ${borValue.toFixed(1)}% dalam rentang optimal`,
        recommendation: 'Pertahankan tingkat utilisasi ini'
      };
    } else {
      return {
        status: 'low',
        level: 'info',
        message: `BOR ${borValue.toFixed(1)}% rendah`,
        recommendation: 'Evaluasi strategi pemasaran/rujukan'
      };
    }
  }

  evaluateLOS(losValue: number): EvaluationResult {
    const los = this.standards.LOS;
    
    if (losValue >= (los.critical_high || 12)) {
      return {
        status: 'critical',
        level: 'danger',
        message: `LOS ${losValue.toFixed(1)} hari terlalu panjang`,
        recommendation: 'Review protokol discharge dan case management'
      };
    } else if (losValue > los.optimal_max) {
      return {
        status: 'warning',
        level: 'warning',
        message: `LOS ${losValue.toFixed(1)} hari di atas optimal`,
        recommendation: 'Evaluasi efisiensi perawatan'
      };
    } else if (losValue >= los.optimal_min) {
      return {
        status: 'optimal',
        level: 'success',
        message: `LOS ${losValue.toFixed(1)} hari dalam rentang optimal`,
        recommendation: 'Pertahankan kualitas perawatan'
      };
    } else {
      return {
        status: 'low',
        level: 'info',
        message: `LOS ${losValue.toFixed(1)} hari pendek`,
        recommendation: 'Monitor kualitas outcome pasien'
      };
    }
  }

  evaluateBTO(btoValue: number): EvaluationResult {
    const bto = this.standards.BTO;
    
    if (btoValue >= (bto.warning_high || 60)) {
      return {
        status: 'high',
        level: 'warning',
        message: `BTO ${btoValue.toFixed(1)} sangat tinggi`,
        recommendation: 'Evaluasi kualitas perawatan dan patient satisfaction'
      };
    } else if (btoValue >= bto.optimal_min) {
      return {
        status: 'optimal',
        level: 'success',
        message: `BTO ${btoValue.toFixed(1)} dalam rentang optimal`,
        recommendation: 'Pertahankan efisiensi turnover'
      };
    } else {
      return {
        status: 'low',
        level: 'info',
        message: `BTO ${btoValue.toFixed(1)} rendah`,
        recommendation: 'Evaluasi strategi admission dan discharge'
      };
    }
  }

  getStatusColor(level: EvaluationLevel): string {
    switch (level) {
      case 'success': return 'text-green-600 bg-green-50 border-green-200';
      case 'warning': return 'text-yellow-600 bg-yellow-50 border-yellow-200';
      case 'danger': return 'text-red-600 bg-red-50 border-red-200';
      case 'info': return 'text-blue-600 bg-blue-50 border-blue-200';
      default: return 'text-gray-600 bg-gray-50 border-gray-200';
    }
  }

  getStatusIcon(level: EvaluationLevel): string {
    switch (level) {
      case 'success': return '✅';
      case 'warning': return '⚠️';
      case 'danger': return '🚨';
      case 'info': return 'ℹ️';
      default: return '❓';
    }
  }
}

// Singleton instance untuk use di seluruh app
export const medicalStandardsEvaluator = new MedicalStandardsEvaluator();

// Hook untuk fetch standards dari API dan update evaluator
export const useMedicalStandards = () => {
  const fetchStandards = async () => {
    try {
      const API_URL = (import.meta as any).env?.VITE_API_URL || 'http://localhost:8000/api/v1';
      const response = await fetch(`${API_URL}/standards/medical`);
      if (response.ok) {
        const result = await response.json();
        medicalStandardsEvaluator.updateStandards(result.data);
        return result.data;
      }
    } catch (error) {
      console.warn('Failed to fetch medical standards from API, using defaults:', error);
    }
    return DEFAULT_MEDICAL_STANDARDS;
  };

  return { fetchStandards };
};
